aip/0160.md

```diff
@@ -0,0 +1,167 @@
+confusion.
+
+### Negation Operators
+
+Filtering implementations **must** provide the unary operators `NOT` and `-`.
+These are used interchangeably.
+
+| Operator | Example | Meaning                  |
+| -------- | ------- | ------------------------ |
+| `NOT`    | `NOT a` | True if `a` is not true. |
+| `-`      | `-a`    | True if `a` is not true. |
+
+### Comparison Operators
+
+Filtering implementations **must** provide the binary comparison operators `=`,
+`!=`, `<`, `>`, `<=`, and `>=`.
+
+| Operator | Example      | Meaning                                         |
+| -------- | ------------ | ----------------------------------------------- |
+| `=`      | `a = true`   | True if `a` is true.                            |
+| `!=`     | `a != 42`    | True unless `a` equals 42.                      |
+| `<`      | `a < 42`     | True if `a` is a numeric value below 42.        |
+| `>`      | `a > "foo"`  | True if `a` is lexically ordered after "foo".   |
+| `<=`     | `a <= "foo"` | True if `a` is "foo" or lexically before it.    |
+| `>=`     | `a >= 42`    | True if `a` is a numeric value of 42 or higher. |
+
+Because filters are accepted as querty strings, type conversion takes place to
+translate the string to the appropriate strongly-typed value:
+
+- Enums expect the enum's string representation (case-sensitive).
+- Booleans expect `true` and `false` literal values.
+- Numbers expect the standard integer or float representations. For floats,
+  exponents are supported (e.g. `2.997e9`).
+- [Durations][] expect a numeric representation followed by an `s` suffix (for
+  seconds). Examples: `20s`, `1.2s`.
+- [Timestamps][] expect an [RFC-3339][] formatted string (e.g.
+  `2012-04-21T11:30:00-0400`). Timezones are supported.
+
+**Warning:** The identifiers `true`, `false`, and `null` only carry intrinsic
+meaning when used in the context of a typed field reference.
+
+Additionally, when comparing strings for equality, the `*` character denotes
+wildcard syntax; for example, `a = "*.foo"` is true if `a` _ends with_ ".foo".
+
+### Traversal operator
+
+Filtering implementations **must** provide the `.` operator, which indicates
+traversal through a message, map, or struct.
+
+| Example         | Meaning                                               |
+| --------------- | ----------------------------------------------------- |
+| `a.b = true`    | True if `a` has a boolean `b` field that is true.     |
+| `a.b > 42`      | True if `a` has a numeric `b` field that is above 42. |
+| `a.b.c = "foo"` | True if `a.b` has a string `c` field that is "foo".   |
+
+**Important:** The `.` operator **must not** be used to traverse through a
+repeated field or list, except for specific use with the `:` operator.
+
+### Has Operator
+
+Filtering implementations **must** provide the `:` operator, which means "has".
+It is usable with collections (repeated fields or maps) as well as messages,
```

Comment:
@TristonianJones For the most part, I downplayed that this also works with Struct and ListValue. It is great that it does, but I expect most users do not know the distinction exists, and will happily just try what they think should work (and find that it does).

Reply:
SGTM, as long as the list and maps are covered appropriately.